### ISSUE TITLE
Loop Control (DEF-2606)

### ISSUE THREAD

**Issue by britzl** [defold/defold#3514]
Description:
Define whether a sound will loop continuously when triggered until stopped, or just a one shot. By default loop should be off.
 Also provide functionality to define whether a sound should loop forever, or just loop a set number of times before stopping.

Use case:

Business need:
2017-04-13 09:24:08 (patrik.nordlander)
The sound component has a loop checkbox, but not an option for number of loops.

**Comment by britzl** [comment]
What is the use case for controlling the number of loops? When is this needed?

**Comment by otsakir** [comment]
This is about how to build finer control over looping, namely determining the number of times a sound should be played. Following the discussion on slack  and after going through the source code, i'll try to describe where we stand and how to move forward. 

Currently, the existing "looping" field is stored as an integer in the .sound file and is treated as a boolean. So, it works like this:

1 : loop forever
0 : no looping - play once

Ideally, we could use the same field to convey all looping options:  

-1 : loop forever
0 or missing :  play once
1 : still play once
\>1 : play many times

The problem here is that this would break existing projects where loop is 'true' i.e. 1. Sounds that would be looping forever will now be played only once.

After taking into account @JCash comments and while trying to avoid regression issues, i'm proposing the following:

#### Editor UI 

New field named 'numloops' in editor. Add a textbox that would either be empty of hold integer values. We could restrict the user to give values >= 2. The whole thing, from UX standpoint,  will work like this:

| Looping       | NumLoops     | Operation    |
--- | --- | ---
|  unchecked (0) or missing | any value or empty   | play once    |
| checked (1)   | <= 0 or empty | loop forever |
| checked (1)   | 1 | play once |
| checked (1)   | n >1 | play n times |

I guess this will be implemented as a separate issue.

#### Resource 

At the resource layer (*.sound project files and  dmGameSystem::Sound data structure) a new field named 'numloops' will be added. It will be effective only if "Looping" is true (i.e. 1). 

#### Sound component

No changes in dmGameSystem::SoundComponent

#### dmSound::SoundInstance and logic

Hand tight now :-) . There is some sort of state to handle here. We need to store the number of remaining loops somewhere. We can use m_Looping field. It will no longer work as a flag but will contain the number of remaining iterations. When a new "play sound" is initiated, it will get initialized from the resource. Each time the sound ends playing, it will get decreased. 

It will work like this:

-1 : loop forever
0 : looping disabled. When the sound ends playing no loop will occur
\>0 : number of loops to do

The field will be initialized here based on resource `looping` and `numloops`:

https://github.com/defold/defold/blob/dev/engine/gamesys/src/gamesys/components/comp_sound.cpp#L389

After each loop SoundInstance.m_Looping will be decreased. Until it reaches zero. This can be handled here:

https://github.com/defold/defold/blob/dev/engine/sound/src/sound.cpp#L1155

@britzl , @JCash  wdyt ?  Is it worth the fuss ?

**Comment by britzl** [comment]
Excellent work @otsakir. I'll let @JCash have a look and give feedback.

**Comment by JCash** [comment]
Nice overview of the issue @otsakir !

For the SoundInstance, I think keeping the boolean (1 bit) as it is would suffice. We just have to add a counter for number of loops. Is a `uint8_t` enough, or do we need a `uint16_t`? Seems a bit unlikely having more that 255 loops is necessary?
(We can always increase later, but it's much harder to go back to 8 bits)

Regarding the editor UI, I'd like it done within the scope of this issue.

All in all, I think it's a good introduction task to the engine, and I think it might be a good addition.

**Comment by otsakir** [comment]
Thanks @JCash .

> For the SoundInstance, I think keeping the boolean (1 bit) as it is would suffice.

+1  that makes things simpler and safer  

> Is a uint8_t enough, or do we need a uint16_t? Seems a bit unlikely having more that 255 loops is necessary?

I also pondered this. I agree. A unsigned byte seems long enough and scales too.

> Regarding the editor UI, I'd like it done within the scope of this issue.

Ok. I'll take my chances there. A good intro doc on adding stuff to the editor would certainly help. If you don't have one, i 'll try to create it on the way :-)

**Comment by otsakir** [comment]
@JCash , tell me your opinion...

I'm trying to freeze the logic that governs the SoundInstance looping. I have the existing `looping` boolean field and an additional one named `loopcount` (or `loopcountdown` or something similar) that helps count the number of loops and stop looping when done.

* For no looping, `looping` ==false, `loopcount` is ignored (it is already set to 0)
* For looping N times, `looping`==true, `loopcount` starts at N and counts down to 0. That is, when the sound stops, it will have been played N times.
* But what about looping forever? `loopcount` is a uint8_t. Unsigned stuff to get the 0-255 range. So, we don't have '-1' as an option :-/

I've thought about the following alternatives:

1. Go up to a signed int16 for `loopcount` and support up to 32k iterations and '-1' as special value too.
2. Stick to 8bits but make it signed. Fewer iterations, up to 127. '-1' special value possible.
3. 'transpose' the loopcount semantics by +1. It will will iterate from 255 down to 1. Actually counting 254 iterations. '0' is reserved for the 'loop forever' case. 

I'd take (2) since i still believe that 127 sound iterations will be ok the use cases 255 are ok. Or (1) if size if not a problem. (3) is hackish but it entered my mind and thus was reported above.

Thoughts ?

**Comment by otsakir** [comment]
Moving on with option (2)

**Comment by britzl** [comment]
@otsakir Sorry for not getting back to you! I think 2 is a good choice!

**Comment by otsakir** [comment]
Btw, i think i could use some help on sound_null.cpp file. Where is this used ? What is its purpose ?

I've altered the interface of dmSound::SetLoooping() in sound.cpp but it's also implemented in sound_null.cpp so it get me thinking...

**Comment by britzl** [comment]
> Btw, i think i could use some help with sound_null.cpp file. Where is this used ? What is its purpose ?

It is used when there is no sound device (headless builds I think) or the developer have chosen to remove sound support from the engine (using an app manifest).

**Comment by otsakir** [comment]
I've wrapped the implementation and did some manual tests. 

I'll create automated tests next.
